File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/annotation/Email.java
package com.wesleyelliott.kubwa.annotation;

import com.wesleyelliott.kubwa.rule.EmailRule;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.ANNOTATION_TYPE;
import static java.lang.annotation.ElementType.CONSTRUCTOR;
import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.ElementType.PARAMETER;
import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * Created by wesley on 2016/07/28.
 */
@ValidateUsing(EmailRule.class)
@Retention(RUNTIME)
@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER , TYPE})
public @interface Email {
    int errorMessage();
    String name() default "emailError";

    @Target(TYPE)
    @Retention(RUNTIME)
    @Documented
    public @interface List {
        Email[] value();
    }

}

File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/annotation/Email.kt
package com.wesleyelliott.kubwa.annotation

import com.wesleyelliott.kubwa.rule.EmailRule

/**
 * Created by wesley on 2016/07/28.
 */
@ValidateUsing(EmailRule::class)
@Retention(AnnotationRetention.RUNTIME)
@Target(
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.FIELD,
    AnnotationTarget.ANNOTATION_CLASS,
    AnnotationTarget.CONSTRUCTOR,
    AnnotationTarget.VALUE_PARAMETER,
    AnnotationTarget.CLASS,
    AnnotationTarget.FILE
)
annotation class Email(val errorMessage: Int, val name: String = "emailError") {

    @Target(AnnotationTarget.CLASS, AnnotationTarget.FILE)
    @Retention(AnnotationRetention.RUNTIME)
    @MustBeDocumented
    annotation class List(vararg val value: Email)

}
